### DISCUSSION TITLE
Poor GPU performance for morphing rover

### DISCUSSION THREAD

**Discussion by dietmarwo** [esa/SpOC2#20] [Morphing Rovers Q&A] [answered]
On my machine (AMD 5950 / Nvidia 3090) I was able to reconfigure the morphing rover UDP so that it runs successfully on the GPU. But unfortunately it becomes slower than on CPU. And strangely a second machine with a much weaker 1660 TI
is faster than my 3090 (same CPU). 

- Do I miss something? Is there some Pytorch configuration option improving performance on GPU? Have others a similar experience?
- Are the NNs too small for GPU learning?
- Do you think https://github.com/google/evojax could be an alternative implementation platform more suitable for GPUs? Here at least for the samples given I noticed a dramatic speedup using a big GPU. But maybe morphing rover doesn't fit there for some reason.

**Comment by pinouche** [comment]
I guess it also depends on your batch size? if you do regular step-by-step, GPU probably isn't worth the overhead of passing data from CPU to GPU?

**Reply by dietmarwo** [reply]
>GPU probably isn't worth the overhead of passing data from CPU to GPU

Thats exactly the problem  https://github.com/google/evojax or more generally https://github.com/google/jax
wants to solve by providing a GPU/TPU based reimplementation of numpy. This way CPU-GPU/TPU data transfers
can mostly be avoided. With Pytorch you always will have this issue for smaller sized NNs. Kind of cool that you 
can write GPU/TPU code in JAX via Python annotations. And this not only works for UDPs - EvoJAX even provides
optimization algorithms executed directly on the GPU. See also https://github.com/RobertTLange/evosax . 

With "batch size" you mean the number of maps/scenarios handled by the UDP? This issue is independent of that.
I even experimented with 16 bit floats which the 3090 supports - speedup is minimal loosing too much accuracy for the morphing task.

**Reply by pinouche** [reply]
Nice, thanks for sharing the library, could be useful for this problem. 

Yes, I mean if you could have a larger batch size (like the number of scenarios), then maybe it is worth using the GPU over the CPU. I have not tested using the GPU myself, yet. I may consider it once I start optimizing.

**Comment by DreamTlaloc** [comment, marked as answer]
The challenge was not designed with any particular hardware in mind, and we can't give any advice on how to improve the performance. But we welcome people to try any approach and to discuss different methods for optimisation!

**Reply by dietmarwo** [reply]
>But we welcome people to try any approach and to discuss different methods for optimisation!

Regarding optimization methods you could try evolutionary methods. 
https://github.com/google/evojax/tree/main/scripts/benchmarks provides some hints what algos work well with NNs. 
These results may provide insights independent from the evojax specific implementation of these algos. Since the page is not updated for some time, you also may check more recent PRs linked here https://github.com/google/evojax/tree/main/evojax/algo . These PRs include a description of the methods and more recent benchmark results. Efficient implementations of the best algos may be found in other libraries independent from the JAX framework.
